### ISSUE TITLE
regex fails to build when pulled in via lalrpop-util 0.19.10

### ISSUE THREAD

**Issue by jonhoo** [lalrpop/lalrpop#775]
```
$ cargo new foo
$ cd foo
$ cargo add lalrpop-util --features regex
    Updating crates.io index
      Adding lalrpop-util v0.19.10 to dependencies.
             Features:
             + regex
             + std
             - lexer
$ cargo check
  Downloaded regex v1.8.1
  Downloaded regex-syntax v0.7.1
  Downloaded 2 crates (588.1 KB) in 0.20s
    Checking regex-syntax v0.7.1
    Checking regex v1.8.1
error: `std` feature is currently required to build this crate
   --> /home/jongje/.cargo/registry/src/index.crates.io-6f17d22bba15001f/regex-1.8.1/src/lib.rs:648:1
    |
648 | compile_error!("`std` feature is currently required to build this crate");
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0432]: unresolved import `crate::Error`
  --> /home/jongje/.cargo/registry/src/index.crates.io-6f17d22bba15001f/regex-1.8.1/src/compile.rs:16:5
   |
16 | use crate::Error;
   |     ^^^^^^^^^^^^ no `Error` in the root
   |
help: a similar name exists in the module
   |
16 | use crate::error;
   |            ~~~~~
help: consider importing one of these items instead
   |
16 | use core::error::Error;
   |     ~~~~~~~~~~~~~~~~~~
16 | use core::fmt::Error;
   |     ~~~~~~~~~~~~~~~~
16 | use crate::compile::fmt::Error;
   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~
16 | use crate::compile::hir::Error;
   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~
     and 7 other candidates

For more information about this error, try `rustc --explain E0432`.
error: could not compile `regex` due to 2 previous errors
```

This happens because the dependency on `regex` specifies `default-features = false`:
https://github.com/lalrpop/lalrpop/blob/22aff1bfb3723eceb5632ab86d8f59fdcfd40636/Cargo.toml#L20

But that also disables the `std` feature, which is currently _required_ to build `regex`:
https://github.com/rust-lang/regex/blob/1.8.1/src/lib.rs#L647-L648

If nothing else in the dependency graph takes a dependency on `regex` without `default-features = false` (or enabling `std`), then `regex` fails to build as a result.

This doesn't happen to `lalrpop` (without `-util`), since it explicitly always adds the `std` feature:
https://github.com/lalrpop/lalrpop/blob/22aff1bfb3723eceb5632ab86d8f59fdcfd40636/lalrpop/Cargo.toml#L28

**Comment by yannham** [comment]
I heard about the very same issue on a unrelated repository. The stance from there was to never specify features at the crate-level for workspace-level dependencies. That is, the definition at the workspace level should define all the needed features, and crates like `lalrpop-util` or `lalrpop` should just use them or not as a toggle, but not enable specific features (otherwise, what happened here is a risk, that a feature being enabled or not is a non-local property and depends on the combination of each crate of the workspace).

**Comment by Pat-Lafon** [comment]
Hmm, this is not something I'm too familiar with. If I understand https://doc.rust-lang.org/cargo/reference/resolver.html#features correctly, at the workspace level is unions together all the feature flags of it's members? Does this suggest not having workspace level dependencies at all? Or would it be too wild to set all feature flags from the workspace level.

**Comment by yannham** [comment]
I think your understanding of the union is correct. 

> Or would it be too wild to set all feature flags from the workspace level.

My suggestion was precisely to follow this discipline: enable the features you need at the workspace level. Doing so, the enabled features never depend on one particular crate being modified, added or removed: you can see them by just looking at the workspace `Cargo.toml`. In return, you have the inconvenient of being less granular. But it seemed ok in practice, and worth the price.

**Comment by Pat-Lafon** [comment]
Oh, perfect, that was what I was thinking as well.

**Comment by youknowone** [comment]
I just yanked 0.19.10 and released 0.19.11
Could you check if the problem is fixed?

https://crates.io/crates/lalrpop/0.19.11

**Comment by jonhoo** [comment]
That seems to be working fine as far as I can tell!
